### ISSUE TITLE
Change Request: Display config object name

### ISSUE THREAD

**Issue by RobinTail** [eslint/eslint#19017]
### ESLint version

v9.12.0

### What problem do you want to solve?

The [documentation](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-objects) says:

> **name** - A name for the configuration object. This is used in error messages

But in fact it does not. Regardless of the `--format`, the `name` of the configuration entry is not present in the ESLint output.

### Reproduction

```js
// eslint.config.js
export default [
  {
    name: "source/all", // <——
    files: ["**/*.js"],
    rules: {
      "no-unused-vars": "error",
    },
  },
];
```

```js
// test.js
var foo = "bar";
```

```sh
$ yarn eslint

.../test.js
  1:5  error  'foo' is assigned a value but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

same for `--format json` and `json-with-metadata` — the property is completely uninvolved, despite the statement in the ESLint documentation.

### What do you think is the correct solution?

I'd like the property `name` of the flat config entry:
- to be used by ESLint,
- somehow reported along with the violated rule names in the error messages,
- at least as an opt-in feature with some CLI flag, like `--debug` or `--format verbose`

It MAY look like this:

```sh
$ yarn eslint

.../test.js
  1:5  error  'foo' is assigned a value but never used  no-unused-vars (source/all)

✖ 1 problem (1 error, 0 warnings)
```

### Participation

- [x] I am willing to submit a pull request for this change.

### Additional comments

I believe that informing the user on the config object `name` could be very helpful and insightful when ESLint config has a lot of entries, especially overlapping ones. It can improve clarity on causes for the reported errors. 
Consider a flat config, declaring rules for all sources in one entry, but exceptions for certain source files in another one.

![multiple flat config entries named](https://github.com/user-attachments/assets/3c22eb93-4e31-4520-8f5a-d4f194942310)

**Comment by RobinTail** [comment]
Found out so far that `name` is only processed by `@eslint/eslintrc` though the call chain `extractConfig / createConfig`, that makes an instance of `ExtractedConfig`, having `configNameOfNoInlineConfig`:

https://github.com/eslint/eslintrc/blob/fc9837d3c4eb6c8c97bd5594fb72ea95b6e32ab8/lib/config-array/config-array.js#L283-L287

So that it's only utilized for inline config to display a message like this, where `configNameOfNoInlineConfig` goes into parentheses:

```
'/*globals*/' has no effect because you have 'noInlineConfig' setting in your config (.eslintrc.yml » eslint-config-foo).
```

The `name` property in the regular scenario (flat config entry) is ignored.

**Comment by mdjermanovic** [comment]
> The [documentation](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-objects) says:
> 
> > **name** - A name for the configuration object. This is used in error messages

"error messages" refers to errors thrown by ESLint's config validators, not lint messages. For example, if you have a config object like this:

```js
{
    name: "my-config",
    linterOptions: {
        reportUnusedDisableDirectives: "no"
    }
}
```

ESLint will throw an error with a message that includes the config object's  `name`:

```
Oops! Something went wrong! :(

ESLint: 9.12.0

ConfigError: Config "my-config": Key "linterOptions": Key "reportUnusedDisableDirectives": Expected one of: "error", "warn", "off", 0, 1, 2, or a boolean.
```

> I believe that informing the user on the config object `name` could be very helpful and insightful when ESLint config has a lot of entries, especially overlapping ones. It can improve clarity on causes for the reported errors.

Did you check ESLint Config Inspector? It provides a list of all enabled rules and names of config objects where each rule is enabled.

![Image](https://github.com/user-attachments/assets/ec63fd19-037e-48c3-b314-d2508bd75e0c)


https://eslint.org/docs/latest/use/command-line-interface#--inspect-config

**Comment by RobinTail** [comment]
Thank you for clarifications, @mdjermanovic !

> Did you check ESLint Config Inspector? It provides a list of all enabled rules and names of config objects where each rule is enabled.

Yeah, I checked it. It is a very powerful and comprehensive tool, I think.
But I found its interface a bit complicated when it comes to figuring out where exactly in my configuration the particular rule is enabled on particular file.

Here the user story (as per my experience needed for that):

1. Find out the CLI feature `--inspect-config` and run it
2. Accept installation of an additional package `@eslint/config-inspector`
3. Then it opens a browser and I see a lot of stuff happening there (realize that I'm looking at config visualization)
4. Figure out to click on "Files"
5. Then find my file in a long list (probably use browser search, no search feature available on that tab)
6. Click on the file and get another list — the config objects applicable to that file
7. I won't open each object to find my the particular rule, so I should figure out to click on "Merged rules"
8. Then I get the list of rules enabled on this file
9. Then I click on the problematic rule I forgot where I set it
10. And finally in a pop-up I get the answer "Set to `error` in {name} the 15th config item"

Quite tricky path, I'd say.
But I admit that it's possible that maybe I'm using it wrong, maybe there is a shortcut, or something.

> not lint messages

What would you say if ESLint could immediately provide the hint regarding the config name along with a lint message?
So that all those 10 steps above would not be needed, because it could report this way:

```elm
{line}:{column} {message} {ruleName} ({configName})
```

Wouldn't it be a better experience?

**Comment by mdjermanovic** [comment]
When you open ESLint Config Inspector, you can click on Rules to see all the rules:

![Image](https://github.com/user-attachments/assets/b73d8380-569c-4da6-9667-56b65a5e7a2f)

> `{line}:{column} {message} {ruleName} ({configName})`
>
> Wouldn't it be a better experience?
> Users could at a glance backtrace the reported problems to their config entries! How cool is that? :)

What is the use case where this feature would be particularly helpful?

If the user has no control over the ESLint config in the project they're working on, then the config name is unnecessary information that can just confuse them. If the user has control over the ESLint config, then this info might be useful when adding a shareable configuration, but I believe the user would rather want to see all the rules that the configuration enables than just those that report errors in the current state of the codebase. For this use case, the ESLint Config Inspector seems like a much better tool.

**Comment by RobinTail** [comment]
> What is the use case where this feature would be particularly helpful?

The most difficult part of my job is explaining the real value of my brilliant ideas :)
Here we go, @mdjermanovic 

### First case

Consider the following ESLint config file. It's based on recommended ways of installing ESLint, TypeScript-ESLint and Prettier from the corresponding documentations:

```js
import jsPlugin from "@eslint/js";
import tsPlugin from "typescript-eslint";
import prettierOverrides from "eslint-config-prettier";
import prettierRules from "eslint-plugin-prettier/recommended";

export default [
  jsPlugin.configs.recommended,
  ...tsPlugin.configs.recommended,
  prettierOverrides,
  prettierRules,
];
```

As you might notice, it consists of presets, predefined collection of options and rules, that:
1. are not revealed in this file explicitly,
2. are sometimes arrays (_tsPlugin.configs.recommended_)
3. overlap: disable a rule enabled by previous entry, or replace it with another one, so that the order matters

And then you run `eslint` and get a message that you believe does not fit your desires:

```elm
{row}:{col} error  {message}  {ruleName}
```

How would you know why this particular rule is enabled? Where exactly is it enabled among those presets?
You could run `--inspect-config` and browse UI, **OR** ESLint could just immediately tell you that in CLI without extra effort.

### Second case

Imagine I have test files in the same directory as the sources, like `file.js` and `file.spec.js` (quite popular setup).
And I need stricter ESLint rules for the sources, but more relaxed for the tests, so I override this way:

```js
export default [
  {
    name: "my/sources",
    files: ["src/*.js"],
    rules: {
      "some/rule": "error"
    },
  },
  {
    name: "my/tests",
    files: ["src/*.spec.js"],
    rules: {
      "some/rule": "off"
    }
  }
];
```

All good until I make another test `file.test.js` because I forget to follow the naming convention of my tests (_.test.js_ instead of _.spec.js_). Now I'm getting an error message and trying to figure out why. The hint in the error message could help me immediately realize that my `file.test.js` considered as a source file, not a test:

```elm
{row}:{col} error  {message}  some/rule (my/sources)
```

And this is just a simplified case, because real life configs sometimes look much more complicated and overlapping.

**Comment by mdjermanovic** [comment]
Thanks for the clarifications.

As for the first use case, I still think the ESLint Config Inspector is a much better tool, for the reasons described in https://github.com/eslint/eslint/issues/19017#issuecomment-2412102529.

As for the second use case, this feature indeed could occasionally be helpful to figure out the problem quickly, but this use case is also covered by the ESLint Config Inspector: you can enter a file path to see all config objects that apply to that file.

Given that the use cases are already covered by ESLint Config Inspector, and that this feature would require significant changes to the core (linting a file currently operates with the final calculated config for the file, which doesn't include any info about config objects that were used to calculate it) that could impact the performance, I'm not in favor or this change.

@eslint/eslint-team any other thoughts about this?

**Comment by RobinTail** [comment]
>  covered by ESLint Config Inspector

Of course, everything is covered by Inspector. My point is that:
 
- it's too massive for answering simple questions,
- it's detached from the reported problem (runs in browser where you need to apply extra effort),
- it's a separate package to install.
